kie-issues#1182: Performing the `undo` operation after resetting the expressions crashes the new DMN Editor

danielzhe opened the pull request:
Closes: https://github.com/apache/incubator-kie-issues/issues/1182

The issue was that BEE was providing expressions without IDs, which was not expected by the DMN Editor.

This change guarantees that every expression and its nested components (like variables, context entries, and so on) that have IDs will have its ID defined (no more undefined IDs if they are optional in the XML definitions).

tiagobento commented on packages/dmn-editor/src/mutations/updateExpression.ts:
```diff
@@ -64,4 +65,12 @@ export function updateExpression({
   } else {
     throw new Error("DMN MUTATION: Can't update expression for drgElement that is not a Decision or a BKM.");
   }
+
+  getNewDmnIdRandomizer()
+    .ack({
+      json: definitions.drgElement,
+      type: "DMN15__tDefinitions",
+      attr: "drgElement",
+    })
+    .randomize();
```
This is not necessary with the ids being added in all places we need, right?

danielzhe replied:
Yes, it is probably a leftover from some tests we were doing.

tiagobento commented on packages/boxed-expression-component/src/expressions/ExpressionDefinitionRoot/ExpressionDefinitionLogicTypeSelector.tsx:
```diff
@@ -147,9 +147,7 @@ export function ExpressionDefinitionLogicTypeSelector({
-        return (
-          <DecisionTableExpression expression={expression} isNested={isNested} parentElementId={parentElementId} />
-        );
+        return <DecisionTableExpression expression={expression} isNested={isNested} />;
```
Why remove the parentElementId?

danielzhe replied:
It is not used anywhere. ESLint warned me about so I removed it. 😄

tiagobento commented:
Thanks for the PR @danielzhe! Definitely a much needed improvement. Can you please add a regression test too?

danielzhe commented:
I added some new tests and new stories to expressions that we don't have stories yet. We need more work on that, but we have the base now. I need your review here, @ljmotta .

danielzhe commented:
@tiagobento Changes applied! I added tests to every expression known to break due to the lack of the ID.

P.S.: sorry for the rebase. New PRs are coming only with merge. 😄

ljmotta commented on packages/boxed-expression-component/tests-e2e/__fixtures__/jsonModel.ts:
```diff
@@ -18,6 +18,7 @@
  */
 
 import { Page } from "@playwright/test";
+import { BoxedContext, BoxedEvery, BoxedFilter, BoxedFor, BoxedSome, Normalized } from "../../src/api";
```
We mustn't couple the `src` code with the `test` code, as the test should know it. For this case, we duplicate the code to test folder. I'm not sure where we should put them, maybe with the new API?

danielzhe replied:
Yeah, I think the best place is the new API. I'll push the changes and let me know what you think.

ljmotta reviewed (changes requested):
@danielzhe Thanks for this PR. This will be a huge improvement. You got many missing `ids`! I've made some comments inline, more related to the tests. I guess we still have a leftover from Tiago's review.

danielzhe commented on packages/boxed-expression-component/tests-e2e/boxedExpressions/some/someExpression.spec.ts:
```diff
@@ -0,0 +1,48 @@
+  test.beforeEach(async ({ stories }) => {
+    await stories.openBoxedSome("base");
+  });
+  test("should keep IDs after resetting entries", async ({ bee, jsonModel, page }) => {
```
It doesn't need to be the same, just needs to be there. 
The issue was that the ID was removed when the expression was reset and its ID was not set again when the new expression was set.

Maybe the test name is misleading. What about "should set IDs after resetting entries"?

danielzhe replied:
Also, maybe "entries" is not the right name, only for "context". I'm reviewing it and pushing with new suggestions, let me know what you think!

danielzhe commented:
@ljmotta I removed the leftovers now. I removed it before when @tiagobento spotted it but I guess I did bad things with rebase lol.
I also moved the types to "types.ts" inside the API and changed the test names. I think they are more clear now reflecting what was the real issue and what we're testing.

ljmotta commented on packages/dmn-editor/src/mutations/updateExpression.ts:
```diff
@@ -24,6 +24,7 @@ import {
 } from "@kie-tools/dmn-marshaller/dist/schemas/dmn-1_5/ts-gen/types";
 import { renameDrgElement } from "./renameNode";
 import { Normalized } from "../normalization/normalize";
+import { getNewDmnIdRandomizer } from "../idRandomizer/dmnIdRandomizer";
```
Leftover?

danielzhe replied:
You didn't miss it, I missed it :shyface 
It was a new leftover after I removed the `getNewDmnIdRandomizer`.

ljmotta reviewed (commented):
@danielzhe Thanks for addressing the change request. On my first review I missed one thing that I guess it's a leftover.
